module/extension/Director/src/Director.cpp

```diff
@@ -0,0 +1,254 @@
+     * When it updates we check to see if this would now satisfy the conditions and if it does we emit
+     */
+    struct StateUpdate {
+        StateUpdate(const uint64_t& provider_id_, const std::type_index& type_, const int& state_)
+            : provider_id(provider_id_), type(type_), state(state_) {}
+
+        /// The reaction id of the Provider which was waiting on this type
+        uint64_t provider_id;
+        /// The enum type that this enum was listening for
+        std::type_index type;
+        /// The new value for the enum
+        int state;
+    };
+
+    /**
+     * Adds a Provider to the list of active Providers
+     *
+     * @param data_type the data type that this Provider services
+     * @param r         the reaction that is used to run this Provider
+     * @param action    the action type for this Provider (entering, leaving, normal)
+     * @throws std::runtime_error if `r` has previously had a Provider added
+     */
+    void Director::add_provider(const ProvidesReaction& provides) {
+
+        auto& group = groups[provides.type];
+
+        // Check if we already inserted this element as a Provider
+        if (providers.count(provides.reaction->id) != 0) {
+            throw std::runtime_error("You cannot have multiple Provider DSL words in a single on statement.");
+        }
+
+        auto provider = std::make_shared<Provider>(provides.classification, provides.type, provides.reaction);
+        group.providers.emplace_back(provider);
+        providers.emplace(provides.reaction->id, provider);
+    }
+
+    /**
+     * Removes a Provider from our list when it is unbound
+     *
+     * @param id the reaction id for the Provider to be removed
+     *
+     * @throws std::runtime_error if the Provider which the id refers to has not been loaded
+     */
+    void Director::remove_provider(const uint64_t& id) {
+
+        // If we can find it, erase it
+        auto it = providers.find(id);
+        if (it != providers.end()) {
+
+            // Get the relevant elements
+            auto provider = it->second;
+            auto& group   = groups[provider->type];
+
+            // Unbind any when state monitors
+            for (auto& when : provider->when) {
+                when.handle.unbind();
+            }
+
+            // Erase the Provider from this group
+            auto g_it = std::find(group.providers.begin(), group.providers.end(), provider);
+            if (g_it != group.providers.end()) {
+                group.providers.erase(g_it);
+            }
+
+            // Now we need to deal with the cases where this Providers was in use when it was unbound
+            if (provider->active) {
+                if (group.providers.empty()) {
+                    // This is now an error, there are no Providers to service the task
+                    log<NUClear::ERROR>("The last Provider for a type was removed while there were still tasks for it");
```

Comment:
If you're worried about throwing in this because of destructors, then should you reconsider the throw at the end of this function?

Reply:
Nah in that case I want it to explode.
That line should be unreachable unless something very bad and strange has happened.